### ISSUE TITLE
Week 6: Postgres

### ISSUE THREAD

**Issue by bradreeder** [foundersandcoders/master-reference#32]
This issue is intended for usage by mentor's who get assigned to postgreSQL week and for discussion around the week. As a single source for S-G-C notes and FAC9 feedback, discussion, external feedback, and to commit their work against for review.

**Comment by bradreeder** [comment]
FAC9 feedback #18 :

* Workshop on SQL queries rather than doing SQL zoo.

**Comment by bradreeder** [comment]
Week 6 S-G-C notes:

##### Stop

* Open-ended projects. Concentrating on what to build distracts time away from learning objectives.
* SQL Zoo (good for quizzing, but find a better resource.)
* Workshops need to provide more context.

##### Go 

* More silent time when people are researching READMEs.
* Share weekly SGC notes with upcoming week leaders so feedback is carried forwards.
* More workshops on CSS interspersed through back-end weeks.
* code-reviews on projects from people with specific knowledge like css.
* Provide contextual information on topic at start of week.
* More workshops/ code-along/ good workshops to intro db, less talking about them.

##### Continue

* Concise readme presentations (strong approval)
* Mini stop-go-continues (team retrospectives)
* Meditation lessons.
* Keeping the space tidy.

**Comment by emilyb7** [comment]
Things to cover:

- intro to databases (similar to what we did in FAC9, but with more structure)
- install postgres
- intro to SQL commands (probably our own workshop)

part 1:
-- queries
-- joins

part 2:
-- creating a database, adding data to a database

part 3: (next day probably, preceded by walkthrough)
-- connecting database to node server
-- posting data to database, getting data from database, displaying data (no templating, so need to be creative...)

Research topics:

- elephant sql
- build scripts
- testing
- script injections / safety issues


Project:

- 


Problems:

- no hapi, no templating :'(

**Comment by emilyb7** [comment]
Suggested week structure:

Day one
====

**Comment by emilyb7** [comment]
morning challenges??

**Comment by emilyb7** [comment]
@stevehopkinson @tbtommyb @sohilpandya we would love some feedback!

**Comment by stevehopkinson** [comment]
@emilyb7 Looks like a great start to me. A few thoughts:

1. We need to make sure we give F&C10 a lot more context. Quite a few people in FAC9 didn't come out of the first few days of database week with a clear idea of _why_ you'd use a database, rather than saving data in variables in Node. A higher-level introduction to databases, with an explanation of _why_ we use them, backed up by some real-world examples, would go a long way.

2. There was also a lot of fuzziness about various concepts. It's important that people understand that SQL _isn't a database_ - it's a language for querying databases. It's also important that people understand that PostgreSQL/MySQL/MS-SQL are all quite different, and that you can't necessarily apply things directly from one to the others. It also might be worth communicating the fact that a database is also a server, and that when Node queries Postgres you're essentially making an API call. 

3. Two topics that weren't covered, but would be very useful, are subqueries and database views. Subqueries are _really_ important for doing anything beyond basic queries, and they're not any more complicated than joins. Views also have a lot of value - although if the week is already looking quite full, then this could be dropped.


Is it too late to have another thought about the order in which we're teaching things? Teaching databases before we teach templating feels like it could potentially cause a lot of problems.

**Comment by sohilpandya** [comment]
@emilyb7 @stevehopkinson 🎉 

@stevehopkinson would the whiteboard walkthrough be a good time to talk about all of what you've mentioned in 1 & 2? Potentially give them the low-down once they've had a chance to find the answers themselves on the research day?

With regards to **3** would it be possible to have a morning challenge dedicated to `subqueries` & `database views`?

@emilyb7 great start for the week, have you considered splitting the Workshop in this order?  

**part 1** (DIY) (2 Hours on Monday after the walkthrough/Intro to DBs)
- Learn SQL (potentially using https://www.pgexercises.com/) 

**part 2** (Workshop) (Rest of the time dedicated to the workshop)
- Learn how we use it with Node/Hapi => using `npm pg`
  - setup/load schema locally
  - get/post to DB (simple Queries)
  - Learn to build Queries using PG (TDD style)
  - Setup online DB (heroku postgres for example)
- Very happy to help with the layout of this particular workshop, just grab me at some point if you'd like to talk about the best way to tackle this? @/all

**Potential Research Topics**
- [ ] What are databases and why do we need them? 
- [ ] What is the difference between an SQL and NoSQL database?
- Unsure about Elephant SQL as a research topic, was this one of the topics you had in FAC9?

Generally looking really good! can't wait to see the final shape of this week! 😍

**Comment by bradreeder** [comment]
@emilyb7 @stevehopkinson I remember being taught hapi & templating very late in the course so it can be done, you just have to use DOM manipulation to update the html with the results of the query calls, but I see what you mean. 

I think the hapi week are planning to do a CMS project, so that students can use dbs with templates. So maybe if they do dbs without templates in week 6 it will teach them the pain of what templates provide them, which will be provisioned in the next week? Let me know if you feel this can't be worked around as we'd need to resolve in next few days.

Thanks for raising either way, as those in templates week will need to make sure students know how to use templates & dbs in tandem.

**Comment by emilyb7** [comment]
@bradreeder @stevehopkinson  I think we can work around the Hapi-Database order problems, but it does mean that we will have to make significant changes to the workshops that we did.

Agree with Bradley that feeling the pain a little might be helpful.

The main problem, I'm finding, is fitting all the pre-project stuff into a 2-day slot. We had much longer on our cohort, and there will still fundamental gaps in people's understanding when we went into the projects so I think we really have to up our game in terms of planning!!

@stevehopkinson, great points up there! I think we can fit points one and two into our introductory workshop / whiteboard walk-through (whatever we are calling it).

Agree with @sohilpandya that subqueries would be a good subject for a morning challenge, perhaps on Wednesday? I don't think they're essential to getting to grips with databases, so no need to do them before them.

I've never implemented - or even understood the use case for - SQL views. But you can try and convince me that that's worth covering if you want to :D 

@sohilpandya we didn't do elephant sql specifically, but a more general research topic on hosting databases / node heroku. I think, given the reduced scope of the readmes, that demo-ing how to host a DB with Elephant SQL would be a useful practical task for on group of people, as it relates directly to the project.

I'm unsure about your other two suggestions as I think they're a bit vague and I prefer more practical things. Our proposed workshops for Monday and Tuesday already cover some quite difficult subjects, and if anyone hasn't understood what a database is for and why we use them by Tuesday I think we will have wasted their time. I think this is more something for Monday morning.

@shireenzoaby and I agreed that there's not much point in going into detail in terms of relational and non-relational databases. We don't cover non-relational databases in our week, so let's save it for redis week?

**Comment by tbtommyb** [comment]
This is an excellent start, well done everyone!

I agree with @stevehopkinson that stepping back a little first would be useful. I think it would be super useful to cover [database guarantees](https://en.wikipedia.org/wiki/ACID) and it wouldn't take too long. Presumably part of the Monday workshop as @emilyb7 suggests.

@bradreeder, if the Hapi team are planning to do a CMS, might we do building a CMS as an overarching target, at least for weeks 5-8 (for example)? People would gradually add functionality each week and by the end of that you could have something quite decent.

I agree with @sohilpandya about doing sub-queries and views as morning challenges. If we can get things right in the first few days people should be able to pick it up themselves.

**Comment by nelsonic** [comment]
@emilyb7, @stevehopkinson, @bradreeder, @tbtommyb, are set on a "CMS" project?
or would you be _open_ to a project _suggestion_ that would use a _large_ _public_ dataset and give people a set of **questions** that they would need to answer by:
+ learning how to create a new database 
+ how to _import_ an existing dataset
+ how analyse an existing relational schema 
+ construct a query to answer the question
+ display that data on a chart/graph (_with a bit of D3 or Chart.js_) 😉 

Sorry again that there wasn't a "_complete tutorial_" in time for _your_ F&C cohort... 
[***work***](https://www.youtube.com/watch?v=rtXeXd-VMEM&index=5&list=PLmjpEeBZ-AG9FMxwO7A0wlc5NYZaSXtYO) keeps getting in the way of producing learning materials... 😞 
And I quite like the idea of _other people_ producing workshops/example projects for F&C. ❤️

**Comment by emilyb7** [comment]
@nelsonic i think we're open to all kinds of project ideas, we really struggled to find an idea that wouldn't require authentication or cookies. 

So your idea sounds interesting. Do you have a particular public dataset in mind?

**Comment by tbtommyb** [comment]
@nelsonic yeah, not particularly wedded to a project idea. I just thought that it would be good to have a bit of continuity between weeks. 

Now that I think about it, the DB element of a simple CMS doesn't need to be that complex and they could easily implement it alongside learning Hapi as part of the project for week 7 - thoughts @marisid @msachi @SavageWilliam ?

I really like the idea of having to work from an existing dataset, as it's more real-world. Would we expect them to do basic parsing of a CSV file or similar?

**Comment by nelsonic** [comment]
@tbtommyb good question. I was thinking of just giving people the data as an `db.sql.zip` and giving the instructions to load it on their laptops. probably should make it a _small-ish_ dataset (10Mb) to not overload laptops of people who have Chromebooks, but thinking of making a "_server_" available with a _bigger_ dataset (1Tb) still not "_big data_" but big enough for people to feel the pain of a "_slow query_"... with a view to learning query optimisation...

**Comment by sohilpandya** [comment]
## Project

@nelsonic, is it better for someone who's looking at SQL for the first time to analyse an existing relational database **or** learn how to architect their own relational database by going through the process from scratch? 

I'm of the opinion that if we ask them to create a new db from scratch with 5-7 tables and a few relationships between them. 
  - This would be more manageable for people working on it for the first time than importing a DB that could potentially have a lot more tables and relationships. 
  - Also gets them understand the process behind creating an SQL database(creating schema's etc). 

**cons**
- no big data to churn out data from. 
  - This could be addressed with a 2 hour morning challenge where we ask them to understand a 3rd party db and fetch data from it using a query, potentially optimising it as you refactor it? 

I'd like everyone's thoughts on this. 

Thanks. 


## Testing 

Now I feel we need to incorporate this from the get go. 
Students need to understand (IMO): 
- important of testing db
- how to test/setup db
  - each test should be independent to another 
  - this means db needs to be initiated and flushed before starting each test. 

I had a brief chat with Emily and Shireen and if we are struggling for resources, I'm happy to help.

**Comment by emilyb7** [comment]
New ideas for the project!

Mini small-scale CMS that doesn't need auth. Single page app.

Suggested ideas:
- FAC book sharing system (I can add a book, view all books, reserve a book and unreserve a book)
- Food / coffee recommendations around FAC (I can add a coffee shop, rate it, search existing suggestions)
- FAC events calendar (I can add an event like a meetup to a calendar, view existing events, and sign up to attend -i.e. add my name as there is no auth)

**Comment by stevehopkinson** [comment]
A selection of resources that might work for the workshop element:

- https://www.khanacademy.org/computing/computer-programming/sql
- https://www.codecademy.com/learn/learn-sql (very long)
- https://www.udacity.com/course/intro-to-relational-databases--ud197
- https://www.udemy.com/courses/search/?q=sql&src=ukw&price=price-free&lang=en

At some point, someone (possibly me) will have to go through these and select a good one. Anyone have previous experience with any?

**Comment by sohilpandya** [comment]
Hi @/all, how is the week looking? I know a lot of you will be busy over the next month and I don't want you guys to get in a position that means that we are finalising the details the week before. 

I think it might be a good idea to get together and finalise the structure (it's pretty much there), get the resources/materials finalised and then we can close this issue! 🎉 

Lunch this week work with the 4 of you, I can be present if you wish?

**Comment by tbtommyb** [comment]
Sounds good @sohilpandya . Would towards the end of the week be ok? I'm planning to take some time to go through resources etc this week.

**Comment by sohilpandya** [comment]
@/all did you guys get a chance to meet up and finalize the details/resources? I know you've had a super busy week and it's probably going to be the same this week, but putting some time aside for sorting this out at this stage would also be very helpful! 😄

**Comment by sohilpandya** [comment]
@/all Hope you are all doing well, last week prior to FAC10, time for you guys to reconvene and finalize the details?

I really don't want this to become something that is done at the last minute in week 5.

**Comment by tbtommyb** [comment]
TODO:

Please check Emily's updated schedule. We have the following tasks that need done:

* 1 hour introduction / whiteboard session on databases and SQL (Monday morning).
* Installation workshop - including going through dwyl's [version](https://github.com/dwyl/learn-postgresql) and checking that it's up to date.
* Interactive SQL workshop (Monday afternoon). Following the idea of pgexercises.com but a bit more guided and F&C-specific to make it worth not just doing pgexercises.com at home alone. Good opportunity to drop in new concepts.
* Tuesday morning workshop - following [this](https://github.com/matthewglover/postgres-challenge) but rewritten to not use templates. To give people some starting ideas for the week's project.

I am happy to do the second and can help out / review one of the Monday afternoon / Tuesday workshops as they're a bit more involved.

We will also need someone who can guide people through using node's pg module on Tuesday morning - perhaps whoever does the workshop?

**Completion date**: let's aim for end of Feb to have decent quality drafts of everything ready.

**Comment by tbtommyb** [comment]
Also it would be useful to start thinking about who'll do what when.

If possible I'd prefer to do a couple of days consecutively rather than dropping in and out due to work

**Comment by emilyb7** [comment]
also need a morning challenge for wednesday!

**Comment by emilyb7** [comment]
we haven't allotted any time for external speakers or whatever, should we assume we don't have one?

**Comment by emilyb7** [comment]
i wouldn't mind having a go at Tuesday morning's workshop. However I think i will need help! It will be good prep for our Anna Freud project anyway...

**Comment by shiryz** [comment]
I can work on the monday afternoon workshop and the morning challenge as well.

**Comment by stevehopkinson** [comment]
I'm happy to do the intro and afternoon workshop - would it make sense to have one person doing both to keep things consistent through the first day?

**Comment by shiryz** [comment]
@stevehopkinson, I was thinking to work on the workshop because I don't see how I can help @emilyb7 remotely. 
Emily, do you think it'll be worth giving it a go? we can try doing it each on our own and then joining forces :smile: .

**Comment by emilyb7** [comment]
@shiryz, if you can think of a way to split it up then yes sure! I don't
have time today, so if you want to make a start I will pick it up next week.

**Comment by emilyb7** [comment]
Fyi not going to be in London that week anyways. Lucy is covering!

**Comment by tbtommyb** [comment]
I have put initials against the tasks. I will make a repo for the Monday afternoon workshop so that we can get started on it (my assigned tasks are a bit light so happy to help out).

**Comment by tbtommyb** [comment]
See [here](https://github.com/FAC9/postgres-workshop/pull/1) for a skeleton Monday workshop. Feel free to either merge the PR or push more commits to it. I think it might be easier if we keep discussion in this issue though.

In terms of data, I originally thought about doing people/things at F&C but that might be a bit personal and also not useful to new people. Does anyone have any better ideas?

**Comment by bradreeder** [comment]
@emilyb7 @stevehopkinson @tbtommyb @shiryz 

Hey peeps. Good to see your plans up in the PR. 👍 

I've just thought with @emilyb7 and @shiryz in Nazareth, and @stevehopkinson and @tbtommyb both in employment now, I wondered whether you'd all discussed plans for who will be covering the days of this week in London?

If your timetables are likely to be limited, if you could just ensure to have the workshops ready in advance, I can ask around for people in the space who will be available. 👍

**Comment by tbtommyb** [comment]
We were just discussing this on gitter.

So I would like to do Monday and one of the project days (don't really mind much which one).

I think it should be ok if we have just one of us on the project days (possibly plus someone in the space who's free) but we'd need two during the workshop days.

**Comment by stevehopkinson** [comment]
If it's fine with everyone, I'll take Monday and Tuesday – does that work for everyone and match up with the content?

**Comment by bradreeder** [comment]
Week delivered. Thanks @emilyb7 @stevehopkinson @tbtommyb @sohilpandya @shiryz !
